### ISSUE TITLE
[feature request] Simple and Efficient way to get gradients of each element of a sum

### ISSUE THREAD

**Issue by fKunstner** [pytorch/pytorch#7786]
For some application, I need to get gradients for each elements of a sum.
I am aware that this issue has already been raised previously, in various forms ([here](https://discuss.pytorch.org/t/gradient-w-r-t-each-sample/1433/2), [here](https://discuss.pytorch.org/t/efficient-per-example-gradient-computations/17204), [here](https://discuss.pytorch.org/t/quickly-get-individual-gradients-not-sum-of-gradients-of-all-network-outputs/8405) and possibly related to [here](https://github.com/pytorch/pytorch/issues/1407)) 
and has also been raised for other autodifferentiation libraries (some examples for TensorFlow: [here](https://github.com/tensorflow/tensorflow/issues/675), long discussion [here](https://github.com/tensorflow/tensorflow/issues/4897))

While the feature does exists in that there is a way to get the desired output, 
I have not found an efficient way of computing it after investing a significant amount of time.

I am bringing it up again, with some numbers showing the inefficiency of the existing solutions I have been able to find.

It is possible that the running time I observed is not an issue with the existing solutions but only with my implementation of them, 
or that my attempts have been misguided and that a simpler solution exists.
If there is a way to make that operation in an efficient way, I would love to know about it.

I can spend more time working on this issue, but as of now I do not know of a way forward.
The following contains a description of the desired feature, existing "workaround" and some evaluation of their performance.

---

## Feature description:

Given an objective that is a sum of functions, `f(x) = f_1(x) + ... + f_N(x)`, 
a _simple to use_ and _efficient_ way to compute the gradient with respect to `x` for each of the individual function `f_n`, i.e., getting
`[∇f_1(x), ..., ∇f_N(x)]`

If there already is a way to do so, an example would be very nice.

(I understand that this would involve additional memory overhead if `N` is large, as pointed out [here](https://discuss.pytorch.org/t/quickly-get-individual-gradients-not-sum-of-gradients-of-all-network-outputs/8405/2) - my setting is not bound by memory but by time)

## Use case:
- Computing approximate second-order information (Generalized Gauss-Newton type of algorithm).
- Computing gradient statistics such as the variance at a given point.
Some papers based on those ideas: [here](https://pdfs.semanticscholar.org/42e2/1cd78f578fa6ce61b06b99848697da85ed76.pdf), [here](http://www.cs.toronto.edu/~jmartens/docs/Deep_HessianFree.pdf)

## Existing workarounds:
- The `naive` implementation: do the forward pass in batch up to the last part of the function to minimize overhead, and then call backward on each individual `f_n(x)`.
- Goodfellow showed how to recover the individual gradients from gradients with respect to the activation functions in the case of feed-forward neural networks [here](https://arxiv.org/abs/1510.01799). This requires an additional derivation and performs the linear parts of the transformation in Python instead of C, making it scale badly. 
- What I call "Multiple Models": Define copies of the parameters `x_1, ..., x_N` and compute `grad(f_1(x_1) + ... + f_N(x_N))`. This ensures that the gradients are not accumulated, however it scales poorly. Adapted from the priginal formulation [here](https://github.com/tensorflow/tensorflow/issues/4897#issuecomment-290997283)

## Evaluation of existing workarounds
I have tried to make the simplest example of a sufficiently complex problem where this becomes an issue.
The code to reproduce these is available [here](https://github.com/fKunstner/fast-individual-gradients-with-autodiff).

Running on a simple Multi-Layer Perceptron, I am comparing the running time of
* `full`: computing the gradient of the full objective function
* `naive`, `goodf`, `multi`: computing the gradient of the sum by first computing individual gradients with the methods described above, and then taking the sum[1].

I am taking a sum over `N = 1000` elements, and tried different network configurations;
```
# D: Dimension of input and of each Layer
# L: Number of hidden layers
```
For a (relatively) wide and shallow network (`D, L = 500, 1`), I get the following running time
```
Full  :  0.04s
naive : 15.42s
goodf :  4.42s
multi :  1.89s
```
For a narrower and deeper network (`D, L = 100, 10`) I get
```
Full  :  0.03s
naive : 11.40s
goodf :  1.70s
multi :  1.60s
```

While `goodf` and `multi` definitely are improvements on the naive method, they are still ~50 time slower than simply computing the sum. I would not expect any method that stores individual gradients to perform as well as a method that can simply throw them away, but they are essentially doing the same operations so it should be possible to do better.
With those numbers, training a small-ish neural network that would normally take ~10 minutes would take 10 hours, making experimentation very difficult.

From the [information I have been able to find](https://discuss.pytorch.org/t/how-the-hook-works/2222/2), it seems difficult to get access to the un-aggregated gradients.
It is a bit frustrating to know that the backward pass on the function computes the sum in almost no time, and all that is needed for this feature is a way to intercept elements of this sum.

A hook that would allow to store individual gradients on the fly in some list or tensor - maybe with the use of those hooks or by defining a new sum function that intercepts gradients - would be amazing, but I do not understand the codebase and/or AD to do it. 
Some pointers in that direction would already be a big help

[1] _Note: I am not looking for an expensive way of computing the sum by computing individual gradients - I am taking the sum in the end so that all methods do the "same thing", but what is really needed is the intermediate matrix containing individual gradients_

**Comment by ezyang** [comment]
Let `f1x = f_1(x)`, and so forth. Does `torch.autograd.grad((f1x, ..., fnx), (x,))` compute the information you seek?

**Comment by fKunstner** [comment]
@ezyang Thanks for your suggestion!

> Let `f1x = f_1(x)`, and so forth. Does `torch.autograd.grad((f1x, ..., fnx), (x,))` compute the information you seek?

It would have been lovely, but sadly no.
The documentation from `torch.autograd.grad(outputs, ...)` mentions 
> Computes and returns the sum of gradients of outputs w.r.t. the inputs.
> ...

I need the gradient of each element of the sum, not the sum of the gradient of each element.

I [double-checked](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/suggestions.py), to see if it was a documentation problem, and I only get one gradient out of the `.grad` call, not all of them.

**Comment by fKunstner** [comment]
@apaszke Thanks for the detailled answer.

> I don't think that it's possible to reduce the complexity of the standard AD algorithms from being linear in the number of different differentiated functions [...]

I wouldn't expect it to be better than linear either - the issue is not in terms of asymptotical computational complexity but in the overhead involved in doing the aggregation by repeatedly calling backward vs. doing it in one batch in C.

I'd bet this is the reason why `multi` performs better than `naive`.
From my (admittedly limited) understanding of the inner workings, the batch backward computation to leverage vectorization would be ideal.
Even a 2-5x slower than "computing gradients for the sum" implementation would already be a big leap from the 50-400x slowdown experienced using those neat tricks.

> but this would require some pretty significant changes [and] unfortunately I don't think we'll be going down this path anytime soon.

I am a bit sad to read that :(. Trying to work on some algorithms leveraging more information from each batch, I really hoped that I could make my experiments run in minutes instead of hours/hours instead of days :|

While I do not have a strong background in C, could you point me to the file/function/region of the codebase handling the accumulation of the gradients? I don't have hopes of actually implementing it myself, but I am curious as to how this is done (and if maybe I can hack this instead).

As a _bad_ workaround, do you know if this is a supported operations in other Autodifferentiation libraries with GPU support?

Thank you very much for your time.

**Comment by t-vi** [comment]
Really, if you intercept the grad output of a a layer (with backward hook or retain_grad), you get AD "90% of the way" for free (except that you retain the grad).
Now the last step to the individual parameter is more difficult. It's elementary and relatively easy to figure out for linear layers (say in = b x m, weight = m x n, out = b x n, usually grad_weight = in.t() @ out, now you do grad_weight_per_sample = in.unsqueeze(1) * grad_out.unsqueeze(2)). However, for other layers, in particular convolutional layers, that does not work as easily (you could fall back to iterating if you are able to call the backward functions, which you cannot in PyTorch in general).

**Comment by fKunstner** [comment]
@t-vi Thank you for your answer.

If I understand your recommendation correctly, it is very similar to [Goodfellow's derivation](https://arxiv.org/abs/1510.01799) - although my implementation is explicit in asking the gradient for each layer using `torch.autograd.grad` instead of using hooks.

The implementation is still very slow when compared to the computation of the sum of the gradients done by calling `grad` only once - on the order of x50 to x100 slower for a minibatch of size 1000 in my experiments.
Also, assuming the computational complexity scales quadratically in the dimensionality of the network (I might be wrong on that point), doing that part of the differentiation "by hand" in python scales very badly to bigger problems.

My hope was to find a way to leverage the fast implementation.

Do you believe using gradient hooks on the cost function w.r.t. the original parameters would be faster than asking for the gradient of the layers directly?
Even if only 2x faster, that would still be a useful improvement.

**Comment by t-vi** [comment]
If you tag me (tom) on the forums with a pointer to code, we can look there. I think this is beyond the bug.

**Comment by fKunstner** [comment]
I have found _something_ that gives me hope.

By using (probably abusing) the `grad_outputs=None` arguments of `torch.autograd.grad` and passing a `torch.eye(N).view(N,N,1)`, I am able to get individual gradients when using `torch.stack` on simple expressions.
From my limited understanding, using this 3D tensor as a starting point makes the "accumulation" of the different gradients happen in different dimensions.

[Runnable example](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/pytorch/proof_of_concept.py)

```
########################################
# Proof of concept - "Working", somewhat

x1 = torch.tensor([1,1,1], requires_grad=False).float()
x2 = torch.tensor([2,2,2], requires_grad=False).float()
w = torch.tensor([0,0,0], requires_grad=True).float()

f1 = torch.dot(x1, w)
f2 = torch.dot(x2, w)
stacked = torch.stack((f1, f2))

grads, = torch.autograd.grad(stacked, w, torch.eye(2).view(2,2,1), retain_graph=True)

####
# grads contains tensor([[ 1.,  1.,  1.], [ 2.,  2.,  2.]])
# which is the desired output, [∇w f1, ∇w f2]
```

However, this breaks down the backward pass on more involved graphs.
I am probably messing with the intended structure of the matrix multiplications in the backward pass a bit too much.

```
import torch.nn.functional as F

f1 = F.sigmoid(torch.dot(x1, w))
f2 = F.sigmoid(torch.dot(x2, w))
stacked = torch.stack((f1, f2))

grads, = torch.autograd.grad(stacked, w, torch.eye(2).view(2,2,1), retain_graph=True)

####
# output and gradOutput have different number of elements: 
# output[1] has 1 elements, while gradOutput[2 x 1] has 2 elements 
# at c:\programdata\miniconda3\conda-bld\pytorch-cpu_[...]
# \work\aten\src\thnn\generic/Sigmoid.c:19
```

If someone could look at this idea and perhaps give me a few pointer to better understand what is happening under the hood, this would be very much appreciated.

---

@t-vi I did open a [forum thread](https://discuss.pytorch.org/t/efficient-computation-of-per-sample-examples/18587)

**Comment by apaszke** [comment]
Yes, what you did will work for some simple ops, but really is an implementation detail and is an overuse of the system.

Provided you don't use too many operations as part of your model there is a relatively simple workaround that would let you achieve perfect batching, although you'll need to reimplement the derivative for all the ops that you need, and use them instead of the library ones (`nn` won't work either). Here's an example:

```py
import torch

# These custom ops break the usual assumption that gradients given to
# backward have the same shapes as outputs. They are expected to have
# an extra leading dimension, which batches independent reverse mode
# passes.
class BatchedReverseMM(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, y):
        ctx.save_for_backward(x, y)
        return torch.matmul(x, y)

    @staticmethod
    def backward(ctx, batched_grad):
        x, y = ctx.saved_variables
        return torch.matmul(batched_grad, y.t()), torch.matmul(x.t(), batched_grad)

class BatchedReLU(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        ctx.mask = x >= 0
        return torch.relu(x)

    @staticmethod
    def backward(ctx, batched_grad):
        return batched_grad * ctx.mask.type_as(batched_grad).expand_as(batched_grad)

matmul = BatchedReverseMM.apply
relu = BatchedReLU.apply

batch_size, input_size, output_size = 20, 100, 10
x = torch.randn(batch_size, input_size)
W = torch.randn(input_size, output_size, requires_grad=True)

# NOTE: need to use the custom ops here
output = relu(matmul(x, W))
jac_elems = output_size * batch_size # it's really the size of one dim of the jacobian
batch_grad_output = torch.eye(jac_elems, jac_elems).view(jac_elems, *output.shape)

jacobian, = torch.autograd.grad(output, W, batch_grad_output)
print(jacobian.shape)
```

**Comment by fKunstner** [comment]
@apaszke Excellent example, it covers the basics I need.
I will experiment with this and report some numbers asap.

Many many thanks for the helping hand, it is very much appreciated.

**Comment by fKunstner** [comment]
Status update - playing a bit with @apaszke's example.

[Test-code](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/tree/master/pytorch) updated, though it is becoming messy.
I'll try to clean it up with some examples when it's done.
* [Batch gradient ops](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/pytorch/batchFuncs.py)
* [Model class](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/pytorch/models.py)
* [Testing script](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/pytorch/main.py) (messy)

An obvious-in-retrospect issue with the current implementation is that the backward operations involve `[N x N x 1]` tensors multiplication, and thus scales in O(N^2).

---

The following table illustrates the issue with various batch sizes.
The model is a MLP with sigmoid activations, 10-dimensional inputs and two hidden layers of 10 nodes.

N=    | 10 | 100 | 1000 | 5000
--    | -- | -- | -- | --
Full  | 0.000 | 0.003 | 0.008 | 0.060
stack | 0.008 | 0.005 | 0.243 | 4.350
naive | 0.008 | 0.050 | 0.595 | 7.118
goodf* | 0.000 | 0.010 | 0.087 | 0.303
multi | 0.005 | 0.074 | 0.397 | 1.803

(in seconds, truncated to %.3f)

*This benchmark oversells the performance of this method which has a much higher constant factor on the `O(D^2)` scaling.

---

This however is not a limitation of the idea, only the current implementation using multiplication with those huge tensors. In every one of those multiplication, one of the dimension is useless - in that only a single line of each dimension is nnz.

I think a more careful implementation could scale - for example by only computing product for nnz lines or resizing "on the fly" instead of storing those huge matrices.

@t-vi I only now realise (through your [forum post](https://discuss.pytorch.org/t/efficient-computation-of-per-sample-examples/18587/2)) that your [earlier post](https://github.com/pytorch/pytorch/issues/7786#issuecomment-391327871) was using the same trick. I will try to play around with the re-definition of backward for those ops and see if I can make this work in the next few days (though feel free to ping in if the solution seems obvious :)

Thanks again!

**Comment by Mercurial1101** [comment]
@fKunstner I have tried your code and also run with my own examples. It seems that in your implementation, good fellow's algorithm runs even slower than the naive implementation when the dimension of layer becomes large. It is quite weird. Could you comment on it?

**Comment by fKunstner** [comment]
@Mercurial1101 I'm glad this thread is relevant to other people!

I have some thoughts about what might be happening, but they might very well be wrong. 
My todolist includes revisiting this approach, I might have a better answer after more digging but here are some ideas if you want to try something now.

---

My understanding of AD is poor and the implementation of Goodfellow's trick might be the issue, or the adaptation of the trick to retrieve individual gradients, not the trick in itself.

The derivation in [his note](https://arxiv.org/pdf/1510.01799.pdf) is concerned with the specific case of retrieving the L2 norm of the gradient, and it is possible that the intermediate gradients retrieved by the backward pass are optimized for this use. It is possible this information to reconstruct the individual gradients, but it involves re-doing part of the backward pass, which is what is done in the file [`goodfellow_backprop.py`](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/blob/master/pytorch/goodfellow_backprop.py).

The backward pass is supposed to scale in `O(P^2)`, where `P` is the number of node in the hidden layers, since it involves `P x P` matrix multiplications, but the way it is done might put a higher-than-necessary constant in front.

Possible not-mutually-exclusive reasons:
* The current code might be doing a very bad job at matrix multiplication by doing things serially in the for loop instead of using a bigger matrix operation that would leverage parallelization.
* The intermediate gradients being asked might be a good starting point to construct the L2 norm of the gradient, instead of the gradient; asking for the right starting matrix might be the issue.
* The method still adds overhead in that it needs to retrieve and store 2 big matrices per layer and multiply them, which has "already been done" by the backward pass, only with a reduction step over the minibatch dimension.

**Comment by fKunstner** [comment]
Conclusion:
* If you use `bmm` instead of computing outer products in a for loop, Goodfellow's trick actually performs well.
* It turns out to be the same as what @t-vi proposed [here](https://github.com/pytorch/pytorch/issues/7786#issuecomment-391327871) and is equivalent to redefining the derivative of the linear model, as @apaszke showed how to do [here](https://github.com/pytorch/pytorch/issues/7786#issuecomment-391637797) - though doing this outside of the backward pass code might be a bit cleaner.

For a reasonably big model (50 layers, 300 hidden unit/layer) änd minibatch size (~100), it performs only ~10x worse than computing the full gradient, and ~10x better than computing individual gradients by repeatedly calling `backward`. 

I have some numbers/setups [here](https://raw.githubusercontent.com/fKunstner/fast-individual-gradients-with-autodiff/master/pytorch/results.txt) and the code to reproduce those results on your machine/setup if you wish to do so is [here](https://github.com/fKunstner/fast-individual-gradients-with-autodiff/tree/master/pytorch), with a (hopefully correct now) version of goodfellow's trick.

I also tried to explain how the trick works in the [readme](https://github.com/fKunstner/fast-individual-gradients-with-autodiff), but I don't think I did a very good job. Hope it's still useful to other people who end up in this situation in the future though.

---

Many thanks to @apaszke and @t-vi!
I really appreciated the helping hand and learnt a lot about AD in the past few days :D 
I'm sorry I took your time on something that eventually ended up being an implementation mistake on my part.

**Comment by Mercurial1101** [comment]
I have checked your code and I think it is correct for linear feedward NN. I am currently considering to do this for NN with more complex structures, like AlexNet, ResNet, etc.

**Comment by singlasahil14** [comment]
Is there a way to find gradient of individual samples in a batch using pytorch multiprocessing module?

**Comment by yaroslavvb** [comment]
@Mercurial1101 that version only works for matmul layers, but for convolutions you would need the trick that rewrites convolution as matrix multiplication. IE, it's the "get_gathering_filter" in [here](https://gist.github.com/Thrandis/9b3f75a130ec6c24a64117b7d9304c3f#file-kfac-py-L204) or extract_patches function in TensorFlow KFAC implementation, or the "expansion operator" that is described [here](http://proceedings.mlr.press/v48/grosse16-supp.pdf)

**Comment by soumith** [comment]
@yaroslavvb FYI, getting the gathering filter = constructing the toeplitz matrix, and can be done with `torch.unfold` much more efficiently.

**Comment by yaroslavvb** [comment]
@soumith thanks, unfold indeed did the trick
I added per-example gradient computation to https://github.com/cybertronai/autograd-hacks

**Comment by ndrmnl** [comment]
Hi all, our team at Owkin has done some small benchmarks on this issue, as well as proposed a solution which seems to be faster at least part of the time. A report is available on [arXiv](https://arxiv.org/abs/1912.06015), and some sample code [at this repo](https://github.com/owkin/grad-cnns).

We have cited @fKunstner's experiments, and used the methods in there as baselines. We have *not* discussed the unfold trick as it led to severe slowdown in some tests we did (although we might have missed something).

In summary, we adapt Goodfellow's trick for convolutional layers, and additionally use the `groups` argument in PyTorch's convolution to be able to handle multiple per-sample convolutions at once (essentially be exchanging the batch and input channel dimensions).

Our intention was not to provide a definitive solution but essentially to continue this discussion, while also showing that we found something that seems to work well. Please let us know if you have any questions or comments :)

**Comment by yaroslavvb** [comment]
@ndrmnl I had some feedback but your email is hard to find :) . Nice job making it practical, I did correctness checks on the unfold trick but never checked performance.

BTW, Goodfellow's trick is actually a special case of einsum optimization. For a linear layer you can write vector of per-example gradient norms squared as the following einsum:

```
torch.einsum("ni,ni,nk,nk->n", A, A, B, B)
```

If you stick this expression into `opt_einsum` package, it discovers Goodfellow's expression when using `optimize=dp` setting. If you do similar trick on convolution gradients using unfold trick, you get different optimal aggregation strategies depending on dimensions of the kernel -- one strategy is quadratic in number of patches while another is quadratic in number of channels.

That approach relies on `torch.unfold`, I'd be curious if you get a way to get optimal strategy that depends on kernel size without calling unfold.

BTW, I recently cleaned-up helper lib I use to capture activations/backprops, it's here https://github.com/cybertronai/autograd-lib

**Comment by ndrmnl** [comment]
@yaroslavvb Thanks for the comment! Btw our emails can be found on the top of the repo I linked above :)

Nice to learn about `opt_einsum`, we had no idea about it! Will take a closer look.

For us performance is really fundamental, since we are mainly interested in the DP application, where one needs to bound the norm of the individual gradients. Differentially-private training of CNNs is typically very impractical, the fastest way we found was by using what @fKunstner calls `multi` above (which was introduced by Goodfellow [here](https://github.com/tensorflow/tensorflow/issues/4897)), but even that was too slow sometimes. What we describe in the report takes around as long as `multi` to run, but can be faster sometimes (and slower in other).

We did a few preliminary experiments with `unfold`, and we had the impression that it was quite heavy both in terms of runtime and memory usage. We only did a naive implementation though, which we did not try to optimize at all. Perhaps with your lib it will be faster, we'll try it out!

**Comment by pulkin** [comment]
I kindly ask to help me here: I am new to all this business and I need a fast and efficient automatic differentiation of the NN I constructed.
```python
class Scalar(nn.Module):
    def __init__(self, n_features):
        super().__init__()
        
        self.l1 = nn.Linear(n_features, 15)
        self.t1 = nn.Tanh()
        self.l2 = nn.Linear(15, 15)
        self.t2 = nn.Tanh()
        self.l3 = nn.Linear(15, 1)
    
    def forward(self, x):
        x = self.t1(self.l1(x))
        x = self.t2(self.l2(x))
        x = self.l3(x)
        return x
```
A typical input is `nn(ones([10000, 100, 3]))` producing something of the shape `[1000, 100, 1]`. Now I am going to fit gradients of Scalar wrt 3 features in this case: i.e. 3 numbers instead of one. What is `pytorch` recipe here?

**Comment by albanD** [comment]
Hi, I think you want to check the [tutorial](https://pytorch.org/tutorials/beginner/deep_learning_60min_blitz.html) on the neural network training.

**Comment by pulkin** [comment]
I'd gladly revisit that video but maybe you could give a more specific input?

**Comment by albanD** [comment]
Hi,

To reduce noise on this unrelated issue, let's move this conversation to the forum: https://discuss.pytorch.org/
We use github issues only for bugs or feature requests.
